src/coral_models/asr/ngram_model.py

```diff
@@ -1,81 +1,78 @@
-    except ValueError:
-        dataset = Dataset.from_file(f"{dataset_id}/dataset.arrow")
+    dataset = load_dataset(
+        path=cfg.model.decoder.dataset_id,
+        name=cfg.model.decoder.dataset_subset,
+        split=cfg.model.decoder.dataset_split,
+        use_auth_token=os.getenv("HUGGINGFACE_HUB_TOKEN"),
+    )
+    assert isinstance(dataset, Dataset)
 
-    # Dump dataset to a text file
-    text_path = data_dir / "text_data.txt"
-    with open(text_path, "w") as f:
-        f.write(" ".join(dataset["text"]))
+    # Dump dataset to a temporary text file
+    text_file = tempfile.NamedTemporaryFile(mode="w", delete=False, suffix=".txt")
+    text_file.write(" ".join(dataset["text"]))
 
     # Ensure that the `kenlm` directory exists, and download if otherwise
-    kenlm_dir = Path("kenlm")
+    kenlm_dir = Path.home() / ".cache" / "kenlm"
     if not kenlm_dir.exists():
-        os.system("wget -O - https://kheafield.com/code/kenlm.tar.gz | tar xz")
+        kenlm_url = "https://kheafield.com/code/kenlm.tar.gz"
+        subprocess.Popen(["wget", "-O", "-", kenlm_url], stdout=subprocess.PIPE)
+        subprocess.Popen(["tar", "-xz"], stdin=subprocess.PIPE, stdout=subprocess.PIPE)
+        subprocess.Popen(
+            ["rm", "-rf", "kenlm.tar.gz"],
+            stdin=subprocess.PIPE,
+            stdout=subprocess.PIPE,
+        )
+        subprocess.Popen(["mv", "kenlm", str(kenlm_dir)], stdin=subprocess.PIPE)
 
     # Compile `kenlm` if it hasn't already been compiled
     kenlm_build_dir = kenlm_dir / "build"
     if not kenlm_build_dir.exists():
-        os.system(
-            "mkdir kenlm/build && " "cd kenlm/build && " "cmake .. && " "make -j2"
-        )
+        subprocess.Popen(["mkdir", str(kenlm_build_dir)])
+        subprocess.Popen(["cd", str(kenlm_build_dir)])
+        subprocess.Popen(["cmake", ".."])
+        subprocess.Popen(["make", "-j", "2"])
 
     # Train the n-gram language model if it doesn't already exist
-    correct_ngram_path = models_dir / f"{n}gram.arpa"
+    correct_ngram_path = Path(cfg.model_dir) / f"{cfg.model.decoder.n}gram.arpa"
     if not correct_ngram_path.exists():
-        # If the raw language model does not exist either, then train from
-        # scratch
-        ngram_path = models_dir / f"raw_{n}gram.arpa"
+        # If the raw language model does not exist either, then train from scratch
+        ngram_path = Path(cfg.model_dir) / f"raw_{cfg.model.decoder.n}gram.arpa"
         if not ngram_path.exists():
-            os.system(
-                f"kenlm/build/bin/lmplz -o {n} <" f'"{text_path}" >' f'"{ngram_path}"'
+            subprocess.Popen(
+                [
+                    "kenlm/build/bin/lmplz",
+                    "-o",
+                    cfg.model.decoder.n,
+                    "<",
+                    '"{text_file.name}"',
+                    ">",
+                    '"{ngram_path}"',
+                ]
             )
```

Comment:
I would also avoid `Popen` here. There isn't an issue I think, but `Popen` is pretty low level. Maybe something like
```
            cmd = [
                "kenlm/build/bin/lmplz",
                "-o",
                str(cfg.model.decoder.n),
            ]
            with open(text_file.name, "r") as f_in:
                with open(ngram_path, "w") as f_out:
                    subprocess.run(cmd, stdin=f_in, stdout=f_out)
```

I would probably also open the temporary file in a `with`, or even not use a temp file at all here.

Reply:
@rlrs Changed it now 👍